kymatio/scattering1d/frontend/base_frontend.py

```diff
@@ -373,41 +372,128 @@ def _document(cls):
 
 
 class TimeFrequencyScatteringBase():
-    def __init__(self, FrontendClass, J_fr=None, Q_fr=1, oversampling_fr='auto'):
-        self.FrontendClass = FrontendClass
-        self.J_fr = J_fr
-        self.Q_fr = Q_fr
+    def __init__(self, J_fr=None, Q_fr=1, oversampling_fr='auto',
+                 resample_psi_fr=True, resample_phi_fr=True):
+        self._J_fr = J_fr
+        self._Q_fr = Q_fr
         self.oversampling_fr = oversampling_fr
+        self.resample_psi_fr = resample_psi_fr
+        self.resample_phi_fr = resample_phi_fr
 
     def build(self):
-        # First-order scattering object for the frequency variable
-        self.max_order_fr = 1
-        self.shape_fr = self.get_shape_fr()
-        # Calibrate all filters with respect to the most obtainable freq rows
-        self.N_fr = self.shape_fr[-1]
-        if self.J_fr is None:
-            self.J_fr = self.get_J_fr()
+        self._shape_fr = self.get_shape_fr()
+        max_order_fr = 1
+        if self._J_fr is None:
+            self._J_fr = int(math.log2(self.Q) + 1)
+
+        self.sc_freq = _FrequencyScatteringBase(
+            self._J_fr, self._shape_fr, self._Q_fr, max_order_fr, self.average,
+            self.resample_psi_fr, self.resample_phi_fr, self.oversampling,
+            self.vectorize, self.out_type, self.backend)
+
+    def get_shape_fr(self):
+        """This is equivalent to `len(x)` along frequency, which varies across
+        `psi2`, so we compute for each.
+        """
+        shape_fr = []
+        for n2 in range(len(self.psi2_f)):
+            j2 = self.psi2_f[n2]['j']
+            max_freq_nrows = 0
+            for n1 in range(len(self.psi1_f)):
+                if j2 > self.psi1_f[n1]['j']:
+                    max_freq_nrows += 1
+            shape_fr.append(max_freq_nrows)
+        return shape_fr
+
+    def meta(self):
+        """Get meta information on the transform
+
+        Calls the static method `compute_meta_jtfs()` with the parameters of the
+        transform object.
 
-        self.sc_freq = self.FrontendClass(
-            self.J_fr, self.N_fr, self.Q_fr, self.max_order_fr, self.average,
-            self.oversampling, self.vectorize, self.out_type, self.backend)
+        Returns
+        ------
+        meta : dictionary
+            See the documentation for `compute_meta_jtfs()`.
+        """
+        return compute_meta_jtfs(self.J, self.Q, self.J_fr, self.Q_fr)
+
+    # access key attributes via frequential class
+    @property
+    def J_fr(self):
+        return self.sc_freq.J_fr
+
+    @property
+    def Q_fr(self):
+        return self.sc_freq.Q_fr
+
+    @property
+    def N_fr(self):
+        return self.sc_freq.N_fr
+
+    @property
+    def max_order_fr(self):
+        return self.sc_freq.max_order_fr
+
+    @classmethod
+    def _document(cls):
+        # TODO documentation
+        cls.__doc__ = """Joint time-frequency scattering"""
 
+
+class _FrequencyScatteringBase(ScatteringBase):
```

Comment:
are you sure that there's 0% of the Scattering1DBase code that could be inherited for this?
It's embarassing to see all these constants sigma0, alpha, Pmax, eps being copy-pasted

Reply:
otherwise no big deal